addons/hr/models/hr_employee_departure.py

```diff
@@ -0,0 +1,216 @@
+        for departure in self:
+            if departure.action_at_departure:
+                departure.apply_date = max(departure.departure_date, today)
+            else:
+                departure.apply_date = departure.action_other_date or today
+
+    @api.constrains('employee_id')
+    def _check_departure_validity(self):
+        emps_with_contract_conflict = self.env['hr.employee']
+        emps_with_version_conflict = self.env['hr.employee']
+        for departure in self:
+            active_version_sudo = departure.employee_id.sudo().current_version_id
+            if active_version_sudo.contract_date_start and active_version_sudo.contract_date_start > departure.departure_date:
+                emps_with_contract_conflict += departure.employee_id
+            no_previous_version = not departure.employee_id.version_ids.filtered(
+                lambda v: v.date_version < departure.departure_date,
+            )
+            if no_previous_version:
+                emps_with_version_conflict += departure.employee_id
+        if emps_with_contract_conflict:
+            raise ValidationError(self.env._(
+                "Departure date is earlier than the start date of the current contract(s) of %s.",
+                ', '.join(emps_with_contract_conflict.mapped('name')),
+            ))
+        if emps_with_version_conflict:
+            raise ValidationError(self.env._(
+                "There is no valid version starting before the departure date for %s.",
+                ', '.join(emps_with_version_conflict.mapped('name')),
+            ))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        for departure in res:
+            departure.employee_id._get_version(departure.departure_date).write({
+                'departure_id': departure.id,
+            })
+        return res
+
+    def action_schedule(self):
+        self.state = 'scheduled'
+
+    def action_cancel(self):
+        self.state = 'cancelled'
+        self.employee_id.version_id.departure_id = False
+
+    def _cron_apply_departure(self):
+        departures = self.search([
+            ('state', '=', 'scheduled'),
+            ('apply_date', '<=', fields.Date.today()),
+        ])
+        departures.action_register()
+
+    def action_register(self):
+        def _get_user_archive_notification_action(message, message_type, next_action):
+            return {
+                'type': 'ir.actions.client',
+                'tag': 'display_notification',
+                'params': {
+                    'title': self.env._("User Archive Notification"),
+                    'type': message_type,
+                    'message': message,
+                    'next': next_action,
+                },
+            }
+
+        departures_per_user = self.grouped(lambda departure: departure.employee_id.user_id)
+        to_archive_users = self.env['res.users']
+        unarchived_users = self.env['res.users']
```

Comment:
Nitpicking but to stay consistent with emp_to_archive.

```suggestion
        users_to_archive = self.env['res.users']
        users_to_keep_active = self.env['res.users']
```